### PR TITLE
feat(expressions): Add support for spel functions provided by plugins

### PR DESCRIPTION
**Pull request by robzienert**
- About the simplest plugin interface I could think of. This just changes the expressions code to be aware of `PluginManager` and makes the `ExpressionFunctionProvider` interface extend `ExtensionPoint`.
- I also moved the `ExpressionFunctionProvider` into the correct source set (`src/main/kotlin` instead of `src/main/java`) and added some docs around the interface.
- spotless shifted a bunch of stuff around, so the changeset is larger than what one would expect.

There will be a subsequent PR in Orca that will wire things up.

Example extension point implementation that would get picked up by Orca:

```java
@Extension
public class HelloExpressionFunctionProvider implements ExpressionFunctionProvider {
  @NotNull
  @Override
  public Functions getFunctions() {
    return new Functions(
      new FunctionDefinition("hello")
    );
  }

  @Nullable
  @Override
  public String getNamespace() {
    return null;
  }

  public static String hello() {
    return "hello";
  }
}
```

### CODE REVIEW

File: kork-expressions/src/main/kotlin/com/netflix/spinnaker/kork/expressions/ExpressionFunctionProvider.kt [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ *  }
+ * ```
+ *
+ * The above function provider could then be called in a SpEL expression:
+ *
+ * ```
+ * ${netflix.hello(this)} // `this` is an Execution, if running in Orca.
+ * ```
+ */
+interface ExpressionFunctionProvider : ExtensionPoint {
```
Comment (ajordens) [inline kork-expressions/src/main/kotlin/com/netflix/spinnaker/kork/expressions/ExpressionFunctionProvider.kt]: Thoughts on `@Beta` or `@Alpha` marking extension points when they're first created? 
(I realize this was pre-existing so perhaps not needed in this case)

File: kork-expressions/src/main/kotlin/com/netflix/spinnaker/kork/expressions/ExpressionFunctionProvider.kt [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ *    @JvmStatic
+ *    fun hello(execution: Execution): String =
+ *      "Hello, ${execution.id}"
+ *  }
+ * ```
+ *
+ * The above function provider could then be called in a SpEL expression:
+ *
+ * ```
+ * ${netflix.hello(this)} // `this` is an Execution, if running in Orca.
```
Comment (christopherthielen) [inline kork-expressions/src/main/kotlin/com/netflix/spinnaker/kork/expressions/ExpressionFunctionProvider.kt]: ```// `this` is an Execution, if running in Orca.```
🤯

File: kork-expressions/src/main/kotlin/com/netflix/spinnaker/kork/expressions/ExpressionFunctionProvider.kt (hunk 1 above)
Comment (robzienert) [reply kork-expressions/src/main/kotlin/com/netflix/spinnaker/kork/expressions/ExpressionFunctionProvider.kt]: Yes, my intention is to add the `@Beta` tag. I'm happy to introduce our own `@Alpha` / `@Beta` annotations into `kork-core` if we standardize on something like that.

File: kork-expressions/src/main/kotlin/com/netflix/spinnaker/kork/expressions/ExpressionFunctionProvider.kt (hunk 2 above)
Comment (robzienert) [reply kork-expressions/src/main/kotlin/com/netflix/spinnaker/kork/expressions/ExpressionFunctionProvider.kt]: Yeah. I'm going to remove that. No sense in documenting some fanciness that is Orca-specific in kork.

**Pull request merged**